File: elixir/lib/symphony/agent_runner.ex
defmodule Symphony.AgentRunner do
  @moduledoc """
  Executes the full agent pipeline for a single issue:
  workspace creation -> hooks -> NodeBridge session -> prompt -> result -> cleanup.
  """
  require Logger

  alias Symphony.{Config, Workspace, PromptBuilder}
  alias Symphony.Claude.NodeBridge

  defstruct [
    :issue,
    :workspace_dir,
    :session_id,
    :bridge,
    :tracker,
    :hooks,
    :workflow
  ]

  @default_template ~S"""
  You are working on issue #{{identifier}}: {{title}}

  ## Description
  {{description}}

  ## Instructions
  Implement the requested changes. Create a branch, make commits, and open a PR when done.
  """

  @doc """
  Run the full agent pipeline for an issue. Returns {:ok, result} or {:error, reason}.
  Called from Task.Supervisor.async_nolink in the Orchestrator.
  """
  def run(issue, opts \\ []) do
    bridge = Keyword.get(opts, :bridge, NodeBridge)
    tracker = Keyword.get(opts, :tracker, Symphony.GitHub.Adapter)
    hooks = Keyword.get(opts, :hooks, %{})
    workflow = Keyword.get(opts, :workflow)

    identifier = Symphony.GitHub.Issue.identifier(issue)
    session_id = "symphony-#{identifier}-#{System.unique_integer([:positive])}"

    Logger.info("[AgentRunner] Starting issue ##{identifier}: #{issue.title}")

    broadcast(:agent_started, %{issue_number: identifier, title: issue.title})

    runner = %__MODULE__{
      issue: issue,
      session_id: session_id,
      bridge: bridge,
      tracker: tracker,
      hooks: hooks,
      workflow: workflow
    }

    result = run_pipeline(runner, identifier)

    # Always try to clean up workspace
    if runner.workspace_dir do
      Workspace.remove(runner.workspace_dir, hooks: hooks)
    end

    result
  end

  defp run_pipeline(runner, identifier) do
    with {:ok, dir} <- create_workspace(runner),
         runner = %{runner | workspace_dir: dir},
         :ok <- Workspace.before_run(dir, runner.issue, hooks: runner.hooks),
         {:ok, prompt} <- build_prompt(runner),
         {:ok, result} <- execute_agent(runner, prompt),
         :ok <- post_result(runner, result),
         :ok <- Workspace.after_run(dir, runner.issue, hooks: runner.hooks) do
      broadcast(:agent_completed, %{
        issue_number: identifier,
        result: result.result,
        cost_usd: result.cost_usd,
        num_turns: result.num_turns
      })

      Logger.info("[AgentRunner] Completed issue ##{identifier} (#{result.num_turns} turns, $#{result.cost_usd})")
      {:ok, result}
    else
      {:error, reason} = err ->
        Logger.error("[AgentRunner] Failed issue ##{identifier}: #{inspect(reason)}")
        broadcast(:agent_failed, %{issue_number: identifier, error: inspect(reason)})
        try_comment(runner, "Symphony agent failed: #{inspect(reason)}")
        err
    end
  end

  # --- Pipeline Steps ---

  defp create_workspace(runner) do
    Workspace.create(runner.issue, hooks: runner.hooks)
  end

  defp build_prompt(runner) do
    template =
      if runner.workflow do
        runner.workflow.template
      else
        @default_template
      end

    PromptBuilder.render(template, runner.issue)
  end

  defp execute_agent(runner, prompt) do
    _identifier = Symphony.GitHub.Issue.identifier(runner.issue)

    session_config = %{
      session_id: runner.session_id,
      model: Config.model(),
      permission_mode: Config.permission_mode(),
      allowed_tools: Config.allowed_tools(),
      max_turns: Config.max_turns(),
      workspace_dir: runner.workspace_dir
    }

    with {:ok, _} <- NodeBridge.session_start(runner.bridge, session_config),
         {:ok, result} <- NodeBridge.turn_start(runner.bridge, %{
           session_id: runner.session_id,
           prompt: prompt
         }, timeout: 600_000) do
      NodeBridge.session_stop(runner.bridge, runner.session_id)

      {:ok, %{
        result: result["result"] || "",
        cost_usd: result["cost_usd"] || 0,
        num_turns: result["num_turns"] || 0
      }}
    else
      {:error, reason} ->
        NodeBridge.session_stop(runner.bridge, runner.session_id)
        {:error, {:agent_execution, reason}}
    end
  end

  defp post_result(runner, result) do
    identifier = Symphony.GitHub.Issue.identifier(runner.issue)
    cost = result.cost_usd
    turns = result.num_turns

    body = """
    ## Symphony Agent Result

    **Status:** Completed
    **Turns:** #{turns}
    **Cost:** $#{Float.round(cost * 1.0, 4)}

    ### Output
    #{String.slice(result.result, 0, 60_000)}
    """

    runner.tracker.comment(identifier, body)
  end

  defp try_comment(runner, message) do
    if runner.tracker && runner.issue do
      identifier = Symphony.GitHub.Issue.identifier(runner.issue)
      runner.tracker.comment(identifier, message)
    end
  rescue
    _ -> :ok
  end

  defp broadcast(event, payload) do
    Phoenix.PubSub.broadcast(
      Symphony.PubSub,
      "symphony:events",
      {event, payload}
    )
  end
end

File: src/agent-runner.ts
import { Effect, PubSub } from "effect"
import { Config } from "./config.js"
import { EventBus, type SymphonyEvent } from "./event-bus.js"
import { TrackerService } from "./github/tracker.js"
import { type Issue, identifier } from "./github/issue.js"
import { type Workflow } from "./workflow.js"
import { ensureWorkspace, beforeRun, afterRun, type WorkspaceError } from "./workspace.js"
import { render, type TemplateError } from "./prompt-builder.js"
import { runTurn, type TurnResult } from "./claude-session.js"
import { GhCliError } from "./github/client.js"

const DEFAULT_TEMPLATE = `You are working on issue #{{identifier}}: {{title}}

## Description
{{description}}

## Instructions
Implement the requested changes. Create a branch, make commits,
and open a PR that references this issue.`

export interface AgentResult {
  readonly result: string
  readonly costUsd: number
  readonly numTurns: number
  readonly exitReason: "normal" | "error"
}

/**
 * Run the full agent pipeline for a single issue.
 * Loops through multiple turns (up to max_turns), re-checking issue state
 * after each turn per the spec (Section 7.1 / 16.5).
 */
export function runAgent(
  issue: Issue,
  workflow: Workflow | null,
  hooks: Record<string, string> = {},
  attempt: number | null = null,
): Effect.Effect<AgentResult, WorkspaceError | TemplateError | GhCliError | Error, Config | EventBus | TrackerService> {
  const id = identifier(issue)

  return Effect.gen(function* () {
    const config = yield* Config
    const pubsub = yield* EventBus
    const tracker = yield* TrackerService

    yield* PubSub.publish(pubsub, {
      _tag: "AgentStarted",
      issueNumber: id,
      title: issue.title,
    } satisfies SymphonyEvent)

    // Ensure workspace (persists across runs)
    const ws = yield* ensureWorkspace(issue, hooks, config.hookTimeoutMs)
    yield* beforeRun(ws.path, issue, hooks, config.hookTimeoutMs)

    const template = workflow?.template ?? DEFAULT_TEMPLATE
    const maxTurns = config.maxTurns
    let turnNumber = 1
    let totalCostUsd = 0
    let totalTurns = 0
    let lastResult = ""

    if (config.dryRun) {
      // Simulate agent work without calling Claude
      const delayMs = 2000 + Math.random() * 3000
      yield* Effect.sleep(`${Math.round(delayMs)} millis`)
      totalCostUsd = 0
      totalTurns = 1
      lastResult = `[dry-run] Simulated agent work for issue #${id}: ${issue.title}`
    } else {
      while (turnNumber <= maxTurns) {
        // First turn: full rendered prompt. Continuation turns: guidance only.
        const prompt = turnNumber === 1
          ? yield* render(template, issue, attempt)
          : `Continue working on issue #${id}: ${issue.title}. This is turn ${turnNumber}/${maxTurns}. Check the current state and continue where you left off.`

        const sessionId = `symphony-${id}-${Date.now()}`
        const turnResult: TurnResult = yield* runTurn(
          {
            sessionId,
            model: config.model,
            permissionMode: config.permissionMode,
            allowedTools: config.allowedTools,
            maxTurns: 0, // let Claude CLI decide when done; outer loop controls symphony turns
            workspaceDir: ws.path,
          },
          { sessionId, prompt },
        )

        totalCostUsd += turnResult.costUsd
        totalTurns += turnResult.numTurns
        lastResult = turnResult.result

        // Re-check issue state from tracker after each turn
        const currentIssue = yield* tracker.getIssue(id).pipe(
          Effect.catchAll(() => Effect.succeed(null)),
        )

        // If issue is no longer in an active state, stop
        if (!currentIssue || !isActiveState(currentIssue.state, config.trackerActiveStates)) {
          break
        }

        if (turnNumber >= maxTurns) {
          break
        }

        turnNumber++
      }
    }

    // Best-effort after_run hook
    yield* afterRun(ws.path, issue, hooks, config.hookTimeoutMs)

    // Verify a PR was actually created
    const hasPR = yield* tracker.hasLinkedPR(id).pipe(
      Effect.catchAll(() => Effect.succeed(false)),
    )

    const status = hasPR ? "Completed" : "Completed (no PR found)"

    // Post result comment (best-effort)
    const commentBody = [
      "## Symphony Agent Result",
      "",
      `**Status:** ${status}`,
      `**Turns:** ${totalTurns}`,
      `**Cost:** $${totalCostUsd.toFixed(4)}`,
      ...(hasPR ? [] : ["", "> **Warning:** No linked pull request was found for this issue."]),
      "",
      "### Output",
      lastResult.slice(0, 60_000),
    ].join("\n")

    yield* tracker.comment(id, commentBody).pipe(
      Effect.catchAll(() => Effect.void),
    )

    const result: AgentResult = {
      result: lastResult,
      costUsd: totalCostUsd,
      numTurns: totalTurns,
      exitReason: "normal",
    }

    yield* PubSub.publish(pubsub, {
      _tag: "AgentCompleted",
      issueNumber: id,
      result: result.result,
      costUsd: result.costUsd,
      numTurns: result.numTurns,
    } satisfies SymphonyEvent)

    if (!hasPR) {
      yield* PubSub.publish(pubsub, {
        _tag: "AgentCompletedWithoutPR",
        issueNumber: id,
        costUsd: result.costUsd,
        numTurns: result.numTurns,
      } satisfies SymphonyEvent)
    }

    return result
  }).pipe(
    Effect.tapError((err) =>
      Effect.gen(function* () {
        const pubsub = yield* EventBus
        const tracker = yield* TrackerService
        yield* PubSub.publish(pubsub, {
          _tag: "AgentFailed",
          issueNumber: id,
          error: String(err),
        } satisfies SymphonyEvent)
        // Best-effort error comment
        yield* tracker.comment(id, `Symphony agent failed: ${String(err)}`).pipe(
          Effect.catchAll(() => Effect.void),
        )
      }),
    ),
  )
}

function isActiveState(state: string, activeStates: readonly string[]): boolean {
  const normalized = state.toLowerCase()
  return activeStates.some((s) => s.toLowerCase() === normalized)
}
